js/src/common/Model.ts

```diff
@@ -0,0 +1,383 @@
+
+  /**
+   * Whether or not the resource exists on the server.
+   */
+  exists: boolean = false;
+
+  /**
+   * The data store that this resource should be persisted to.
+   */
+  protected store: Store | null;
+
+  /**
+   * @param data A resource object from the API.
+   * @param store The data store that this model should be persisted to.
+   * @public
+   */
+  constructor(data: ModelData = {}, store = null) {
+    this.data = data;
+    this.store = store;
+  }
+
+  /**
+   * Get the model's ID.
+   *
+   * @final
+   */
+  id(): string | undefined {
+    return 'id' in this.data ? this.data.id : undefined;
+  }
+
+  /**
+   * Get one of the model's attributes.
+   *
+   * @final
+   */
+  attribute<T = unknown>(attribute: string): T {
+    return this.data?.attributes?.[attribute] as T;
+  }
+
+  /**
+   * Merge new data into this model locally.
+   *
+   * @param data A resource object to merge into this model
+   */
+  pushData(data: ModelData | { relationships?: SaveRelationships }): this {
+    if ('id' in data) {
+      (this.data as SavedModelData).id = data.id;
+    }
+
+    if ('type' in data) {
+      this.data.type = data.type;
+    }
+
+    if ('attributes' in data) {
+      Object.assign(this.data.attributes, data.attributes);
+    }
+
+    if ('relationships' in data) {
+      const relationships = this.data.relationships ?? {};
+
+      // For every relationship field, we need to check if we've
+      // been handed a Model instance. If so, we will convert it to a
+      // relationship data object.
+      for (const r in data.relationships) {
+        const relationship = data.relationships[r];
+
+        let identifier: ModelRelationships[string];
+        if (relationship instanceof Model) {
+          identifier = { data: Model.getIdentifier(relationship) };
+        } else if (relationship instanceof Array) {
+          identifier = { data: relationship.map(Model.getIdentifier) };
+        } else {
+          identifier = relationship;
+        }
+
+        data.relationships[r] = identifier;
+        relationships[r] = identifier;
+      }
+
+      this.data.relationships = relationships;
+    }
+
+    // Now that we've updated the data, we can say that the model is fresh.
+    // This is an easy way to invalidate retained subtrees etc.
+    this.freshness = new Date();
+
+    return this;
+  }
+
+  /**
+   * Merge new attributes into this model locally.
+   *
+   * @param attributes The attributes to merge.
+   * @public
+   */
+  pushAttributes(attributes: ModelAttributes) {
+    this.pushData({ attributes });
+  }
+
+  /**
+   * Merge new attributes into this model, both locally and with persistence.
+   *
+   * @param attributes The attributes to save. If a 'relationships' key
+   *     exists, it will be extracted and relationships will also be saved.
+   * @public
+   */
+  save(
+    attributes: SaveAttributes,
+    options: Omit<FlarumRequestOptions<ApiPayloadSingle>, 'url'> & { meta?: any } = {}
```

Comment:
Do we actually use `meta` anywhere? If so, should we formalize our usage of it via an interface somewhere?